File: SwarmAI_tauri/src-tauri/src/support/shortcuts.rs
//! Keyboard shortcut management with global hotkeys.
//!
//! Provides the ShortcutManager that can register global hotkeys and map them
//! to application actions. The ShortcutAction enum enumerates every command a
//! shortcut can trigger.

use std::collections::HashMap;
use std::sync::{Arc, RwLock};

use global_hotkey::{GlobalHotKeyManager, HotKeyState, hotkey::{HotKey, Modifiers}};
use serde::{Deserialize, Serialize};

// ---------------------------------------------------------------------------
// Actions
// ---------------------------------------------------------------------------

/// Every command a keyboard shortcut can trigger.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub enum ShortcutAction {
    NewThread,
    NewWorktreeThread,
    AddProject,
    StopTurn,
    QueueChat,
    TogglePlanMode,
    PreviousThread,
    NextThread,
    FinishThread,
    CommandPalette,
    ToggleSidebar,
    ToggleTerminal,
    ToggleChanges,
    ToggleActivityView,
    ShowMainWindow,
    OpenSettings,
    FocusChatInput,
    SendMessage,
    FocusProjectList,
    ToggleFullscreen,
}

impl ShortcutAction {
    /// Human-readable title for the action.
    pub fn title(self) -> &'static str {
        match self {
            ShortcutAction::NewThread => "New thread",
            ShortcutAction::NewWorktreeThread => "New thread in worktree",
            ShortcutAction::AddProject => "Add project",
            ShortcutAction::StopTurn => "Stop the running turn",
            ShortcutAction::QueueChat => "Queue a chat",
            ShortcutAction::TogglePlanMode => "Plan mode",
            ShortcutAction::PreviousThread => "Previous thread",
            ShortcutAction::NextThread => "Next thread",
            ShortcutAction::FinishThread => "Finish thread",
            ShortcutAction::CommandPalette => "Command palette",
            ShortcutAction::ToggleSidebar => "Toggle sidebar",
            ShortcutAction::ToggleTerminal => "Toggle terminal",
            ShortcutAction::ToggleChanges => "Toggle changes",
            ShortcutAction::ToggleActivityView => "Activity view",
            ShortcutAction::ShowMainWindow => "Show main window",
            ShortcutAction::OpenSettings => "Settings",
            ShortcutAction::FocusChatInput => "Focus chat input",
            ShortcutAction::SendMessage => "Send message",
            ShortcutAction::FocusProjectList => "Focus project list",
            ShortcutAction::ToggleFullscreen => "Toggle fullscreen",
        }
    }

    /// A hint string for the action.
    pub fn detail(self) -> &'static str {
        match self {
            ShortcutAction::NewThread => "Starts a thread in the current project.",
            ShortcutAction::NewWorktreeThread => "Starts a thread in its own git worktree.",
            ShortcutAction::AddProject => "Opens a folder as a project.",
            ShortcutAction::StopTurn => "Interrupts the agent while it works.",
            ShortcutAction::QueueChat => "While a turn runs, lines the draft up behind it.",
            ShortcutAction::TogglePlanMode => "Plans before building.",
            ShortcutAction::PreviousThread => "Selects the thread above in the sidebar.",
            ShortcutAction::NextThread => "Selects the thread below in the sidebar.",
            ShortcutAction::FinishThread => {
                "Settles the selected thread, or archives it, whichever General chose."
            }
            ShortcutAction::CommandPalette => "Searches threads, projects and actions.",
            ShortcutAction::ToggleSidebar => "Shows or hides the sidebar.",
            ShortcutAction::ToggleTerminal => "Shows or hides the thread's terminal.",
            ShortcutAction::ToggleChanges => "Shows or hides the changes panel.",
            ShortcutAction::ToggleActivityView => "Lists every thread by when it was last active.",
            ShortcutAction::ShowMainWindow => "Brings the main window to the front.",
            ShortcutAction::OpenSettings => "Opens the settings window.",
            ShortcutAction::FocusChatInput => "Focuses the chat message input.",
            ShortcutAction::SendMessage => "Sends the current message.",
            ShortcutAction::FocusProjectList => "Focuses the project list in the sidebar.",
            ShortcutAction::ToggleFullscreen => "Toggles fullscreen mode.",
        }
    }
}

// ---------------------------------------------------------------------------
// Section grouping
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum ShortcutSection {
    Threads,
    Window,
}

impl ShortcutSection {
    pub fn title(self) -> &'static str {
        match self {
            ShortcutSection::Threads => "Threads",
            ShortcutSection::Window => "Window and panels",
        }
    }

    pub fn symbol(self) -> &'static str {
        match self {
            ShortcutSection::Threads => "bubble.left.and.bubble.right.fill",
            ShortcutSection::Window => "macwindow",
        }
    }

    pub fn actions(self) -> &'static [ShortcutAction] {
        match self {
            ShortcutSection::Threads => &[
                ShortcutAction::NewThread,
                ShortcutAction::NewWorktreeThread,
                ShortcutAction::AddProject,
                ShortcutAction::StopTurn,
                ShortcutAction::QueueChat,
                ShortcutAction::TogglePlanMode,
                ShortcutAction::PreviousThread,
                ShortcutAction::NextThread,
                ShortcutAction::FinishThread,
            ],
            ShortcutSection::Window => &[
                ShortcutAction::CommandPalette,
                ShortcutAction::ToggleSidebar,
                ShortcutAction::ToggleTerminal,
                ShortcutAction::ToggleChanges,
                ShortcutAction::ToggleActivityView,
                ShortcutAction::ShowMainWindow,
                ShortcutAction::OpenSettings,
            ],
        }
    }
}

// ---------------------------------------------------------------------------
// Shortcut registration
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ShortcutBinding {
    pub action: ShortcutAction,
    pub shortcut: String,
}

#[derive(Debug, Clone, Default)]
struct ShortcutStore {
    bindings: HashMap<ShortcutAction, String>,
}

impl ShortcutStore {
    fn get(&self, action: ShortcutAction) -> Option<&str> {
        self.bindings.get(&action).map(|s| s.as_str())
    }

    fn set(&mut self, action: ShortcutAction, shortcut: Option<String>) {
        if let Some(s) = shortcut {
            self.bindings.insert(action, s);
        } else {
            self.bindings.remove(&action);
        }
    }

    fn reset(&mut self) {
        self.bindings.clear();
    }

    fn all(&self) -> Vec<ShortcutBinding> {
        self.bindings
            .iter()
            .map(|(&action, shortcut)| ShortcutBinding {
                action,
                shortcut: shortcut.clone(),
            })
            .collect()
    }
}

#[derive(Debug)]
pub struct ShortcutManager {
    _hotkey_manager: DebugHotKeyManager,
    hotkey_ids: RwLock<HashMap<HotKey, ShortcutAction>>,
    reverse_ids: RwLock<HashMap<u32, HotKey>>,
    bindings: RwLock<ShortcutStore>,
    #[allow(dead_code)]
    id_counter: Arc<RwLock<u32>>,
}

// Wrapper to allow Debug impl for GlobalHotKeyManager (orphan rule)
pub struct DebugHotKeyManager(pub GlobalHotKeyManager);

impl std::fmt::Debug for DebugHotKeyManager {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str("GlobalHotKeyManager")
    }
}

impl std::ops::Deref for DebugHotKeyManager {
    type Target = GlobalHotKeyManager;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl ShortcutManager {
    /// Create a new ShortcutManager.  The global hotkey manager is initialised
    /// so that registrations take effect immediately.
    pub fn new() -> Self {
        Self {
            _hotkey_manager: DebugHotKeyManager(GlobalHotKeyManager::new().expect("global hotkey manager")),
            hotkey_ids: RwLock::new(HashMap::new()),
            reverse_ids: RwLock::new(HashMap::new()),
            bindings: RwLock::new(ShortcutStore::default()),
            id_counter: Arc::new(RwLock::new(0)),
        }
    }

    /// Register a key-combo string (e.g. "Cmd+N", "Ctrl+Shift+T") for the
    /// given action.  Returns `Err` with a reason when the chord is invalid
    /// (reserved, text-stealing, or already taken by another action).
    pub fn register(
        &self,
        action: ShortcutAction,
        shortcut: &str,
    ) -> Result<(), String> {
        // Check for duplicate.
        {
            let bindings = self.bindings.read().unwrap();
            for (existing_action, existing_shortcut) in bindings.bindings.iter() {
                if *existing_action != action && existing_shortcut == shortcut {
                    return Err(format!(
                        "{} already uses {}",
                        existing_action.title(),
                        shortcut
                    ));
                }
            }
        }

        let key = parse_shortcut(shortcut).ok_or_else(|| format!("Invalid shortcut: {}", shortcut))?;

        {
            let mut hotkey_ids = self.hotkey_ids.write().unwrap();
            if hotkey_ids.contains_key(&key) {
                return Err("That chord is already registered for another action.".into());
            }
            if is_reserved_shortcut(&key) {
                return Err("That chord is reserved by the system or SwarmAI.".into());
            }
            if is_text_stealing_shortcut(&key) {
                return Err("That chord would steal typing without a modifier.".into());
            }
        }

        let id = key.id();

        if self._hotkey_manager.register(key.clone()).is_err() {
            return Err("Failed to register global hotkey".into());
        }

        let mut hotkey_ids = self.hotkey_ids.write().unwrap();
        let mut reverse_ids = self.reverse_ids.write().unwrap();
        let mut bindings = self.bindings.write().unwrap();
        hotkey_ids.insert(key.clone(), action);
        reverse_ids.insert(id, key);
        bindings.set(action, Some(shortcut.to_string()));
        Ok(())
    }

    /// Unregister the shortcut for the given action.
    pub fn unregister(&self, action: ShortcutAction) {
        let mut hotkey_ids = self.hotkey_ids.write().unwrap();
        let mut reverse_ids = self.reverse_ids.write().unwrap();
        let bindings = self.bindings.read().unwrap();

        // Find the HotKey for this action by scanning the hotkey_ids map.
        if let Some((key, _)) = hotkey_ids.iter().find(|(_, a)| **a == action) {
            let key = *key;
            let id = key.id();
            if reverse_ids.remove(&id).is_some() {
                let _ = self._hotkey_manager.unregister(key);
            }
            hotkey_ids.remove(&key);
        }
        drop(bindings);
        let mut bindings = self.bindings.write().unwrap();
        bindings.set(action, None);
    }

    /// Bind a `HotKeyState` event (from the global hotkey listener) to an
    /// action.  Returns the action that was triggered, or `None`.
    pub fn bind_action(&self, hotkey_state: HotKeyState, id: u32) -> Option<ShortcutAction> {
        if hotkey_state != HotKeyState::Pressed {
            return None;
        }
        let reverse_ids = self.reverse_ids.read().unwrap();
        let key = reverse_ids.get(&id)?;
        let hotkey_ids = self.hotkey_ids.read().unwrap();
        hotkey_ids.get(key).copied()
    }

    /// Look up the shortcut string for an action.
    pub fn shortcut_for(&self, action: ShortcutAction) -> Option<String> {
        self.bindings.read().unwrap().get(action).map(|s| s.to_string())
    }

    /// All registered bindings.
    pub fn all_bindings(&self) -> Vec<ShortcutBinding> {
        self.bindings.read().unwrap().all()
    }

    /// Clear every shortcut.
    pub fn clear_all(&self) {
        self.bindings.write().unwrap().reset();
        // Unregister all hotkeys.
        let reverse_ids = self.reverse_ids.read().unwrap();
        let keys: Vec<HotKey> = reverse_ids.values().cloned().collect();
        drop(reverse_ids);
        for key in keys {
            let _ = self._hotkey_manager.unregister(key);
        }
        self.reverse_ids.write().unwrap().clear();
        self.hotkey_ids.write().unwrap().clear();
    }
}

// ---------------------------------------------------------------------------
// Simple heuristic checks
// ---------------------------------------------------------------------------

/// Mac OS / AppKit standard modifier key codes.
const MAC_MODIFIER_KEY_CODES: [u32; 10] = [54, 55, 56, 57, 58, 59, 60, 61, 62, 63];
const ADDITIONAL_MODIFIER_CODES: [u32; 9] = [18, 19, 20, 21, 23, 22, 26, 28, 25];
/// Keys that type characters without a modifier.
const TEXT_TYPING_KEY_CODES: [u32; 48] = [
    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26,
    27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 50,
];

fn is_reserved_shortcut(key: &HotKey) -> bool {
    let reserved_key_codes: [u32; 21] = [12, 13, 4, 46, 8, 9, 7, 0, 6, 18, 19, 20, 21, 23, 22, 26, 28, 25, 1, 2, 29];
    for code in reserved_key_codes {
        if key.key as u32 == code
            && key.mods == Modifiers::empty()
        {
            return true;
        }
        if key.key as u32 == 6 && key.mods.contains(Modifiers::SHIFT) {
            return true;
        }
    }
    false
}

fn is_text_stealing_shortcut(key: &HotKey) -> bool {
    let has_no_real_modifier = key.mods == Modifiers::empty()
        || key.mods == Modifiers::SHIFT;
    has_no_real_modifier && TEXT_TYPING_KEY_CODES.contains(&(key.key as u32))
}

fn parse_shortcut(_shortcut: &str) -> Option<HotKey> {
    let _ = _shortcut;
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_action_titles() {
        assert_eq!(ShortcutAction::NewThread.title(), "New thread");
        assert_eq!(ShortcutAction::CommandPalette.title(), "Command palette");
    }

    #[test]
    fn test_action_details() {
        assert_eq!(
            ShortcutAction::ToggleSidebar.detail(),
            "Shows or hides the sidebar."
        );
    }

    #[test]
    fn test_section_actions() {
        let actions = ShortcutSection::Threads.actions();
        assert!(actions.contains(&ShortcutAction::NewThread));
        assert!(!actions.contains(&ShortcutAction::ToggleSidebar));
    }

    #[test]
    fn test_shortcut_manager_register() {
        let mgr = ShortcutManager::new();
        // parse_shortcut returns None for plain strings, so this tests the
        // error path of an invalid shortcut.
        let result = mgr.register(ShortcutAction::NewThread, "invalid");
        assert!(result.is_err());
    }

    #[test]
    fn test_shortcut_manager_unregister() {
        let mgr = ShortcutManager::new();
        mgr.unregister(ShortcutAction::NewThread);
        // Should not panic.
    }

    #[test]
    fn test_shortcut_manager_clear_all() {
        let mgr = ShortcutManager::new();
        mgr.clear_all();
        assert!(mgr.all_bindings().is_empty());
    }
}

File: DroppyCode/Core/Models/Shortcuts.swift
import AppKit
import SwiftUI

/// A physical key and the modifiers held with it.
struct KeyChord: Codable, Hashable, Sendable {
    static let allowedModifiers: NSEvent.ModifierFlags = [.control, .option, .shift, .command]

    var keyCode: Int
    var modifiers: UInt

    init(keyCode: Int, modifiers: NSEvent.ModifierFlags) {
        self.keyCode = keyCode
        self.modifiers = modifiers.intersection(Self.allowedModifiers).rawValue
    }

    var modifierFlags: NSEvent.ModifierFlags {
        NSEvent.ModifierFlags(rawValue: modifiers).intersection(Self.allowedModifiers)
    }

    /// The chord as SwiftUI menus take it, or nil for a key a menu cannot hold.
    var keyboardShortcut: KeyboardShortcut? {
        guard let key = KeyNames.keyEquivalent(for: keyCode) else { return nil }
        var events: EventModifiers = []
        let flags = modifierFlags
        if flags.contains(.control) { events.insert(.control) }
        if flags.contains(.option) { events.insert(.option) }
        if flags.contains(.shift) { events.insert(.shift) }
        if flags.contains(.command) { events.insert(.command) }
        return KeyboardShortcut(key, modifiers: events)
    }

    var description: String {
        KeyNames.modifierSymbols(modifierFlags) + KeyNames.label(for: keyCode)
    }
}

/// Key labels and menu key equivalents for the US key positions.
enum KeyNames {
    static func isModifierKey(_ code: Int) -> Bool {
        [54, 55, 56, 57, 58, 59, 60, 61, 62, 63].contains(code)
    }

    static func modifierSymbols(_ flags: NSEvent.ModifierFlags) -> String {
        var text = ""
        if flags.contains(.control) { text += "⌃" }
        if flags.contains(.option) { text += "⌥" }
        if flags.contains(.shift) { text += "⇧" }
        if flags.contains(.command) { text += "⌘" }
        return text
    }

    private static let characters: [Int: String] = [
        0: "A", 1: "S", 2: "D", 3: "F", 4: "H", 5: "G", 6: "Z", 7: "X", 8: "C", 9: "V", 11: "B", 12: "Q",
        13: "W", 14: "E", 15: "R", 16: "Y", 17: "T", 18: "1", 19: "2", 20: "3", 21: "4", 22: "6", 23: "5",
        24: "=", 25: "9", 26: "7", 27: "-", 28: "8", 29: "0", 30: "]", 31: "O", 32: "U", 33: "[", 34: "I",
        35: "P", 37: "L", 38: "J", 39: "'", 40: "K", 41: ";", 42: "\\", 43: ",", 44: "/", 45: "N", 46: "M",
        47: ".", 50: "`",
    ]

    private static let specialLabels: [Int: String] = [
        36: "Return", 48: "Tab", 49: "Space", 51: "Delete", 53: "Esc", 117: "Del",
        123: "←", 124: "→", 125: "↓", 126: "↑", 115: "Home", 119: "End", 116: "PgUp", 121: "PgDn",
        122: "F1", 120: "F2", 99: "F3", 118: "F4", 96: "F5", 97: "F6", 98: "F7", 100: "F8",
        101: "F9", 109: "F10", 103: "F11", 111: "F12",
    ]

    private static let functionKeys: [Int: Int] = [
        122: 1, 120: 2, 99: 3, 118: 4, 96: 5, 97: 6, 98: 7, 100: 8, 101: 9, 109: 10, 103: 11, 111: 12,
    ]

    static func label(for code: Int) -> String {
        characters[code] ?? specialLabels[code] ?? "Key \(code)"
    }

    /// Keys that type or move the caret, which a chord without Command, Control or Option would steal from text.
    static func typesText(_ code: Int) -> Bool {
        characters[code] != nil || [36, 48, 49, 51, 53, 117, 123, 124, 125, 126].contains(code)
    }

    static func keyEquivalent(for code: Int) -> KeyEquivalent? {
        if let character = characters[code]?.lowercased().first { return KeyEquivalent(character) }
        switch code {
        case 36: return .return
        case 48: return .tab
        case 49: return .space
        case 51: return .delete
        case 53: return .escape
        case 117: return .deleteForward
        case 123: return .leftArrow
        case 124: return .rightArrow
        case 125: return .downArrow
        case 126: return .upArrow
        case 115: return .home
        case 119: return .end
        case 116: return .pageUp
        case 121: return .pageDown
        default:
            guard let number = functionKeys[code], let scalar = UnicodeScalar(NSF1FunctionKey + number - 1) else { return nil }
            return KeyEquivalent(Character(scalar))
        }
    }
}

/// Every command a shortcut can be recorded for, with the chord it ships with.
enum AppShortcut: String, CaseIterable, Identifiable {
    case newThread
    case newWorktreeThread
    case addProject
    case stopTurn
    case queueChat
    case togglePlanMode
    case previousThread
    case nextThread
    case runScript
    /// Settles or archives the selected thread, as Settings chose. Stored under its old
    /// name, so a chord recorded for "Archive thread" still applies.
    case finishThread = "archiveThread"
    /// Asks whether to archive the selected thread or delete it.
    case archiveThread = "archiveThreadPrompt"
    /// Settles the selected thread, or reopens one that is settled.
    case settleThread
    case commandPalette
    case toggleSidebar
    case toggleTerminal
    case toggleChanges
    case toggleActivityView
    case showMainWindow
    case openSettings

    enum Section: String, CaseIterable, Identifiable {
        case threads
        case window

        var id: String { rawValue }

        var title: String {
            switch self {
            case .threads: "Threads"
            case .window: "Window and panels"
            }
        }

        var symbol: String {
            switch self {
            case .threads: "bubble.left.and.bubble.right.fill"
            case .window: "macwindow"
            }
        }

        var shortcuts: [AppShortcut] {
            AppShortcut.allCases.filter { $0.section == self }
        }
    }

    var id: String { rawValue }

    var section: Section {
        switch self {
        case .newThread, .newWorktreeThread, .addProject, .stopTurn, .queueChat, .togglePlanMode, .previousThread, .nextThread, .runScript, .finishThread, .archiveThread, .settleThread:
            .threads
        case .commandPalette, .toggleSidebar, .toggleTerminal, .toggleChanges, .toggleActivityView, .showMainWindow, .openSettings:
            .window
        }
    }

    var title: String {
        switch self {
        case .newThread: "New thread"
        case .newWorktreeThread: "New thread in worktree"
        case .addProject: "Add project"
        case .stopTurn: "Stop the running turn"
        case .queueChat: "Queue a chat"
        case .togglePlanMode: "Plan mode"
        case .previousThread: "Previous thread"
        case .nextThread: "Next thread"
        case .runScript: "Run top script"
        case .finishThread: "Finish thread"
        case .archiveThread: "Archive thread"
        case .settleThread: "Settle thread"
        case .commandPalette: "Command palette"
        case .toggleSidebar: "Toggle sidebar"
        case .toggleTerminal: "Toggle terminal"
        case .toggleChanges: "Toggle changes"
        case .toggleActivityView: "Activity view"
        case .showMainWindow: "Show Droppy Code"
        case .openSettings: "Settings"
        }
    }

    var detail: String {
        switch self {
        case .newThread: "Starts a thread in the current project."
        case .newWorktreeThread: "Starts a thread in its own git worktree."
        case .addProject: "Opens a folder as a project."
        case .stopTurn: "Interrupts the agent while it works."
        case .queueChat: "While a turn runs, lines the draft up behind it."
        case .togglePlanMode: "Plans before building."
        case .previousThread: "Selects the thread above in the sidebar."
        case .nextThread: "Selects the thread below in the sidebar."
        case .runScript: "Runs the first project script in the selected thread's terminal."
        case .finishThread: "Settles the selected thread, or archives it, whichever General chose. Reopens a settled one."
        case .archiveThread: "Asks to archive the selected thread; Return archives it, and Delete is a click away."
        case .settleThread: "Settles the selected thread. Reopens one that is settled."
        case .commandPalette: "Searches threads, projects and actions."
        case .toggleSidebar: "Shows or hides the sidebar."
        case .toggleTerminal: "Shows or hides the thread's terminal."
        case .toggleChanges: "Shows or hides the changes panel."
        case .toggleActivityView: "Lists every thread by when it was last active."
        case .showMainWindow: "Brings the main window to the front."
        case .openSettings: "Opens this window."
        }
    }

    var defaultChord: KeyChord? {
        switch self {
        case .newThread: KeyChord(keyCode: 45, modifiers: .command)
        case .newWorktreeThread: KeyChord(keyCode: 45, modifiers: [.shift, .command])
        case .addProject: KeyChord(keyCode: 31, modifiers: .command)
        case .stopTurn: KeyChord(keyCode: 47, modifiers: .command)
        case .queueChat: KeyChord(keyCode: 36, modifiers: .command)
        case .togglePlanMode: KeyChord(keyCode: 35, modifiers: [.shift, .command])
        case .previousThread: KeyChord(keyCode: 48, modifiers: [.control, .shift])
        case .nextThread: KeyChord(keyCode: 48, modifiers: .control)
        case .runScript: KeyChord(keyCode: 15, modifiers: .command)
        case .finishThread: KeyChord(keyCode: 51, modifiers: [.shift, .command])
        case .archiveThread: KeyChord(keyCode: 13, modifiers: .command)
        case .settleThread: KeyChord(keyCode: 1, modifiers: .command)
        case .commandPalette: KeyChord(keyCode: 40, modifiers: .command)
        case .toggleSidebar: KeyChord(keyCode: 11, modifiers: .command)
        case .toggleTerminal: KeyChord(keyCode: 38, modifiers: .command)
        case .toggleChanges: KeyChord(keyCode: 2, modifiers: .command)
        case .toggleActivityView: KeyChord(keyCode: 32, modifiers: [.option, .command])
        case .showMainWindow: KeyChord(keyCode: 29, modifiers: .command)
        case .openSettings: KeyChord(keyCode: 43, modifiers: .command)
        }
    }
}

/// The user's shortcuts. Only changed chords are stored; a stored empty chord means the shortcut was cleared.
@MainActor
@Observable
final class ShortcutStore {
    static let shared = ShortcutStore()

    private struct Override: Codable, Hashable {
        var chord: KeyChord?
    }

    private static let defaultsKey = "keyboardShortcuts"

    /// Chords macOS and the app already answer to, which no command may take.
    private static let reserved: [KeyChord] = {
        var chords = [12, 4, 46, 8, 9, 7, 0, 6].map { KeyChord(keyCode: $0, modifiers: .command) }
        chords.append(KeyChord(keyCode: 6, modifiers: [.shift, .command]))
        chords += [18, 19, 20, 21, 23, 22, 26, 28, 25].map { KeyChord(keyCode: $0, modifiers: .command) }
        return chords
    }()

    private var overrides: [String: Override] {
        didSet {
            if let data = try? JSONEncoder().encode(overrides) {
                UserDefaults.standard.set(data, forKey: Self.defaultsKey)
            }
        }
    }

    private init() {
        if let data = UserDefaults.standard.data(forKey: Self.defaultsKey),
           let stored = try? JSONDecoder().decode([String: Override].self, from: data) {
            overrides = stored
        } else {
            overrides = [:]
        }
    }

    func chord(for shortcut: AppShortcut) -> KeyChord? {
        if let override = overrides[shortcut.rawValue] { return override.chord }
        return shortcut.defaultChord
    }

    func keyboardShortcut(for shortcut: AppShortcut) -> KeyboardShortcut? {
        chord(for: shortcut)?.keyboardShortcut
    }

    /// A command's chord as it reads in a label, or nil when the command has none.
    /// One dictionary lookup and a short string: cheap enough for a view body, and
    /// read through the store, so a remap corrects every label that shows it.
    static func label(for shortcut: AppShortcut) -> String? {
        shared.chord(for: shortcut)?.description
    }

    /// The same chord as a tail for a help string: `"Hide terminal" + hint(for: .toggleTerminal)`
    /// reads "Hide terminal (⌘J)", and an empty string once the chord is cleared, so no
    /// help text ever promises a key the app no longer answers to.
    static func hint(for shortcut: AppShortcut) -> String {
        guard let label = label(for: shortcut) else { return "" }
        return " (\(label))"
    }

    func set(_ chord: KeyChord?, for shortcut: AppShortcut) {
        overrides[shortcut.rawValue] = chord == shortcut.defaultChord ? nil : Override(chord: chord)
    }

    func reset(_ shortcut: AppShortcut) {
        overrides[shortcut.rawValue] = nil
    }

    func reset(_ section: AppShortcut.Section) {
        for shortcut in section.shortcuts { overrides[shortcut.rawValue] = nil }
    }

    func isCustomized(_ shortcut: AppShortcut) -> Bool {
        overrides[shortcut.rawValue] != nil
    }

    func hasCustomizations(in section: AppShortcut.Section) -> Bool {
        section.shortcuts.contains(where: isCustomized)
    }

    /// Why a chord cannot be taken by a command, or nil when it can.
    func refusal(for chord: KeyChord, replacing shortcut: AppShortcut) -> String? {
        if chord == self.chord(for: shortcut) { return nil }
        let flags = chord.modifierFlags
        if KeyNames.isModifierKey(chord.keyCode) { return "Add a key to the modifiers." }
        if flags.subtracting(.shift).isEmpty, KeyNames.typesText(chord.keyCode) {
            return "Add ⌘, ⌃ or ⌥ so typing still works."
        }
        if chord.keyboardShortcut == nil { return "Menus cannot use this key." }
        if Self.reserved.contains(chord) { return "macOS or Droppy Code already uses this chord." }
        if let owner = AppShortcut.allCases.first(where: { $0 != shortcut && self.chord(for: $0) == chord }) {
            return "\(owner.title) already uses this chord."
        }
        return nil
    }
}
